File: Examples/test-suite/perl5/director_detect_runme.pl
use strict;
use warnings;
use director_detect;

{
	package MyBar;
	use base 'director_detect::Bar';
	# workaround until attributes work.
	our %val;
	sub new { my $class = shift;
		my $val = @_ ? shift : 2;
		my $self = $class->SUPER::new();
		$val{$self} = $val;
		return $self;
	}
	sub get_value { my($self) = @_;
		$val{$self}++;
		return $val{$self};
	}
	sub get_class { my($self) = @_;
		$val{$self}++;
		return director_detect::A->new();
	}
	sub just_do_it { my($self) = @_;
		$val{$self}++;
	}
	sub clone { my($self) = @_;
		MyBar->new($val{$self});
	}
}

my $b = MyBar->new();

my $f = $b->baseclass();

my $v = $f->get_value();
my $a = $f->get_class();
$f->just_do_it();

my $c = $b->clone();
my $vc = $c->get_value();

if( ($v != 3) or ($MyBar::val{$b} != 5) or ($vc != 6) ) {
	die "Bad virtual detection";
}

File: Examples/test-suite/php/director_detect_runme.php
<?php

require "tests.php";

// No new functions
check::functions(array());
// New classes
check::classes(array('A','Foo','Bar'));
// No new vars
check::globals(array());

class MyBar extends Bar {
  public $val;

  function __construct($val = 2) {
    parent::__construct();
    $this->val = $val;
  }

  function get_value() {
    $this->val = $this->val + 1;
    return $this->val;
  }

  function get_class() {
    $this->val = $this->val + 1;
    return new A();
  }

  function just_do_it() {
    $this->val = $this->val + 1;
  }

  /* clone is a reserved keyword */
  function clone_() {
    return new MyBar($this->val);
  }
}

$b = new MyBar();

$f = $b->baseclass();

$v = $f->get_value();
$a = $f->get_class();
$f->just_do_it();

$c = $b->clone_();
$vc = $c->get_value();

check::equal($v, 3, "f: Bad virtual detection");
check::equal($b->val, 5, "b: Bad virtual detection");
check::equal($vc, 6, "c: Bad virtual detection");

check::done();
